File: hexchat/now_playing.pl
use strict;

use XML::Simple;

Xchat::register("NowPlaying", "1.0");
Xchat::print("NowPlaying script loaded\n");

my $xml = new XML::Simple(SuppressEmpty => '');
my $xmlFile = (-d '/mnt') ? '/mnt/WD20EARX/now_playing.xml' : 'F:\now_playing.xml';

Xchat::hook_command('np', sub {
	my $data = $xml->XMLin($xmlFile);
	my $message = '';
	if (exists $data->{audio}) {
		if ($data->{name}) {
			$message = 'listening to';
			if ($data->{audio}->{disc_number}) {
				if ($data->{audio}->{track_number}) {
					$message = $message . sprintf(' %01d-%02d', $data->{audio}->{disc_number}, $data->{audio}->{track_number});
				}
				else {
					$message = $message . sprintf(' %01d-?', $data->{audio}->{disc_number});
				}
			}
			elsif ($data->{audio}->{track_number}) {
				$message = $message . sprintf(' %02d', $data->{audio}->{track_number});
			}
			$message = $message . " $data->{name}";
			if ($data->{audio}->{album_title}) {
				$message = $message . " ($data->{audio}->{album_title})";
			}
			if ($data->{audio}->{artist}) {
				$message = $message . " by $data->{audio}->{artist}";
			}
			elsif ($data->{audio}->{album_artist}) {
				$message = $message . " by $data->{audio}->{album_artist}";
			}
		}
	}
	elsif (exists $data->{other}) {
		if ($data->{other}->{source_url}) {
			$message = "playing $data->{other}->{source_url}";
		}
	}
	if ($message) {
		Xchat::command("me is $message");
	}
	else {
		Xchat::print("Error in $xmlFile");
	}
	return Xchat::EAT_ALL;
});

File: hexchat/now_playing.py
from __future__ import unicode_literals

import hexchat
import os
import sys
import xml.etree.ElementTree as ET

__module_name__ = str('NowPlaying')
__module_version__ = str('1.0')
__module_description__ = str('')

hexchat.prnt('NowPlaying script loaded')

xml_file = r'/mnt/WD20EARX/now_playing.xml' if os.path.isdir('/mnt') else r'F:\now_playing.xml'

if sys.version_info[0] >= 3:
	FileNotFoundErrorType = FileNotFoundError
else:
	FileNotFoundErrorType = IOError

def nowplaying_callback(word, word_eol, user_data):
	try:
		data = ET.parse(xml_file).getroot()
		
		message = ''
		
		audio = data.find('audio')
		if audio is not None:
			for name in data.findall('name'):
				message = 'listening to'
				
				disc_number = get_child_int_value(audio, 'disc_number')
				total_discs = get_child_int_value(audio, 'total_discs')
				track_number = get_child_int_value(audio, 'track_number')
				
				if disc_number is not None and total_discs is not None and total_discs > 1:
					if track_number is not None:
						message += ' {0:01d}-{1:02d}'.format(disc_number, track_number)
					else:
						message += ' {0:01d}-?'.format(disc_number)
				
				elif track_number is not None:
					message += ' {0:02d}'.format(track_number)
				
				message += ' {0}'.format(name.text)
				
				for album_title in audio.findall('album_title'):
					message += ' ({0})'.format(album_title.text)
					break
				
				artist = audio.find('artist')
				if artist is not None:
					message += ' by {0}'.format(artist.text)
				else:
					for album_artist in audio.findall('album_artist'):
						message += ' {0}'.format(album_artist.text)
						break
				
				break
		
		else:
			for source_url in (other.findall('source_url') for other in data.findall('other')):
				message = 'playing {0}'.format(source_url)
				break
	
	except (FileNotFoundErrorType, ET.ParseError):
		pass
	
	if message:
		command = 'me is {0}'.format(message)
		
		if sys.version_info[0] < 3:
			command = command.encode('utf-8')
		
		hexchat.command(command)
	else:
		hexchat.prnt('Error in {0}'.format(xml_file))
	
	return hexchat.EAT_ALL

def get_child_int_value(node, child_name):
	for child in node.findall(child_name):
		if child.text:
			return int(child.text)
		else:
			return None
	
	return None

hexchat.hook_command('np', nowplaying_callback)
